# 4.10 Coding Etiquette & Excel Reporting

## Part 1
### Question 1: Import data
### Question 2: Consider any security implications that might exist. Address any PII data in the data.
### Question 3: The Instacart officers are interested in comparing customer behavior in different geographic areas. Create a regional segmentation of the data. Determine whether there’s a difference in spending habits between the different U.S. regions. 
### Question 4: The Instacart CFO isn’t interested in customers who don’t generate much revenue for the app. Create an exclusion flag for low-activity customers (customers with less than 5 orders) and exclude them from the data.
### Question 5: The marketing and business strategy units at Instacart want to create more-relevant marketing strategies for different products and are, thus, curious about customer profiling in their database. Create a profiling variable based on age, income, certain goods in the “department_id” column, and number of dependents. 

In [1]:
# Install libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## Question 1:
### Import data

In [2]:
# Create string of location
path = '/Users/aegarcia/Documents/Documents/01-2022 Instacart Basket Analysis/'
# Import Prepared data: orders, products, customers data
df_instacart = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers.pkl'))

## Question 2: 
### Consider any security implications that might exist. Address any PII data in the data.

In [3]:
# Check the names af all the columns
df_instacart.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_label', 'Busy_days_by_orders',
       'Busiest_period_of_day', 'max_order', 'loyalty_flag', 'avg_price',
       'spending_flag', 'frequency_of_orders', 'order_frequency_flag',
       'last_name', 'gender', 'state', 'age', 'date_joined', 'num_dependents',
       'fam_status', 'income', '_merge'],
      dtype='object')

### The first name was deleted before, now I will delete the "last_name' column since it is identifiable information. We will use the 'user_id' to identify our customers. 

In [4]:
# Get a count of Rows and Columns
df_instacart.shape

(32641268, 31)

In [5]:
# Drop 'last_name' & '_merge' Columns
df_instacart = df_instacart.drop(columns = ['last_name', '_merge'])

In [6]:
# Check columns one more time
df_instacart.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_label', 'Busy_days_by_orders',
       'Busiest_period_of_day', 'max_order', 'loyalty_flag', 'avg_price',
       'spending_flag', 'frequency_of_orders', 'order_frequency_flag',
       'gender', 'state', 'age', 'date_joined', 'num_dependents', 'fam_status',
       'income'],
      dtype='object')

In [7]:
# Get a count of Rows and Columns. Revmoved 2 columns, so we have now 29.
df_instacart.shape

(32641268, 29)

In [8]:
# Before continuing, I will change the type of the variables to save some memory and speed up this process.
df_instacart.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32641268 entries, 0 to 32641267
Data columns (total 29 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_time              int64  
 5   days_since_prior_order  float64
 6   product_id              float64
 7   add_to_cart_order       float64
 8   reordered               float64
 9   product_name            object 
 10  aisle_id                float64
 11  department_id           float64
 12  prices                  float64
 13  price_label             object 
 14  Busy_days_by_orders     object 
 15  Busiest_period_of_day   object 
 16  max_order               int64  
 17  loyalty_flag            object 
 18  avg_price               float64
 19  spending_flag           object 
 20  frequency_of_orders     float64
 21  order_frequency_flag    objec

In [9]:
# I will change all the objects into category
df_instacart = df_instacart.astype({'order_id' : 'category',
                            'user_id' : 'category',
                            'product_name' : 'category',
                            'price_label' : 'category',
                            'Busy_days_by_orders' : 'category',
                            'Busiest_period_of_day' : 'category',
                            'loyalty_flag' : 'category',
                            'spending_flag' : 'category',
                            'order_frequency_flag' : 'category',
                            'gender' : 'category',
                            'state' : 'category',
                            'date_joined' : 'category',
                            'fam_status' : 'category'})


In [10]:
# Next, I will change the integers to what is needed for each variable.
df_instacart = df_instacart.astype({'order_number' : 'int16',
                            'orders_day_of_week' : 'int8',
                            'order_time' : 'int8',
                            'max_order' : 'int16',
                            'age' : 'int8',
                            'num_dependents' : 'int8',
                            'income' : 'int32'})

In [11]:
# Check size again and type-changes are ok
df_instacart.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32641268 entries, 0 to 32641267
Data columns (total 29 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                category
 1   user_id                 category
 2   order_number            int16   
 3   orders_day_of_week      int8    
 4   order_time              int8    
 5   days_since_prior_order  float64 
 6   product_id              float64 
 7   add_to_cart_order       float64 
 8   reordered               float64 
 9   product_name            category
 10  aisle_id                float64 
 11  department_id           float64 
 12  prices                  float64 
 13  price_label             category
 14  Busy_days_by_orders     category
 15  Busiest_period_of_day   category
 16  max_order               int16   
 17  loyalty_flag            category
 18  avg_price               float64 
 19  spending_flag           category
 20  frequency_of_orders     float64 
 21  order_

## Nice! it went down from 7.3GB to 3.7GB

## Question 3:
### 3.a) The Instacart officers are interested in comparing customer behavior in different geographic areas. Create a regional segmentation of the data. You’ll need to create a “Region” column based on the “State” column from your customers data set.

In [12]:
df_instacart.loc[(df_instacart['state'] == 'Maine')|(df_instacart['state'] == 'New Hampshire')|(df_instacart['state'] == 'Vermont')|(df_instacart['state'] == 'Massachusetts')|(df_instacart['state'] == 'Rhode Island')|(df_instacart['state'] == 'Connecticut')|(df_instacart['state'] == 'New York')|(df_instacart['state'] == 'Pennsylvania')|(df_instacart['state'] == 'New Jersey'), 'region'] = 'Northeast'
df_instacart.loc[(df_instacart['state'] == 'Wisconsin')|(df_instacart['state'] == 'Michigan')|(df_instacart['state'] == 'Illinois')|(df_instacart['state'] == 'Indiana')|(df_instacart['state'] == 'Ohio')|(df_instacart['state'] == 'North Dakota')|(df_instacart['state'] == 'South Dakota')|(df_instacart['state'] == 'Nebraska')|(df_instacart['state'] == 'Kansas')|(df_instacart['state'] == 'Minnesota')|(df_instacart['state'] == 'Iowa')|(df_instacart['state'] == 'Missouri'), 'region'] = 'Midwest'
df_instacart.loc[(df_instacart['state'] == 'Delaware')|(df_instacart['state'] == 'Maryland')|(df_instacart['state'] == 'District of Columbia')|(df_instacart['state'] == 'Virginia')|(df_instacart['state'] == 'West Virginia')|(df_instacart['state'] == 'North Carolina')|(df_instacart['state'] == 'South Carolina')|(df_instacart['state'] == 'Georgia')|(df_instacart['state'] == 'Florida')|(df_instacart['state'] == 'Kentucky')|(df_instacart['state'] == 'Tennessee')|(df_instacart['state'] == 'Mississippi')|(df_instacart['state'] == 'Alabama')|(df_instacart['state'] == 'Oklahoma')|(df_instacart['state'] == 'Texas')|(df_instacart['state'] == 'Arkansas')|(df_instacart['state'] == 'Louisiana'), 'region'] = 'South'
df_instacart.loc[(df_instacart['state'] == 'Idaho')|(df_instacart['state'] == 'Montana')|(df_instacart['state'] == 'Wyoming')|(df_instacart['state'] == 'Nevada')|(df_instacart['state'] == 'Utah')|(df_instacart['state'] == 'Colorado')|(df_instacart['state'] == 'Arizona')|(df_instacart['state'] == 'New Mexico')|(df_instacart['state'] == 'Alaska')|(df_instacart['state'] == 'Washington')|(df_instacart['state'] == 'Oregon')|(df_instacart['state'] == 'California')|(df_instacart['state'] == 'Hawaii'), 'region'] = 'West'

In [13]:
# Counts on each regions:
df_instacart['region'].value_counts(dropna = False)

South        10870719
West          8353182
Midwest       7652535
Northeast     5764832
Name: region, dtype: int64

In [14]:
# Check counts
10870719+8353182+7652535+5764832

32641268

#### There are no NANs in the new column "Region" and the sum adds usp to the same number of rows of the dataframe

### 3.b) Determine whether there’s a difference in spending habits between the different U.S. regions. (Hint: You can do this by crossing the variable you just created with the spending flag.)

In [15]:
# Create the crosstab
cross_region_spending = pd.crosstab(df_instacart['region'], df_instacart['spending_flag'], dropna = False)
cross_region_spending

spending_flag  High Spender  Low spender
region                                  
Midwest              157347      7495188
Northeast            109140      5655692
South                211679     10659040
West                 161834      8191348

In [16]:
# Copy to Excel
cross_region_spending.to_clipboard()

#### I copied the table  'crosstab' to excel. Then I divided each value of the table by the Table_total = 32641268, which is the number of observations there are in the dataframe. And got this percentages:

#### Region	      High Spender	   Low spender
#### Midwest	    0.48%	            22.96%
#### Northeast	   0.33%	           17.33%
#### South	          0.65%	              32.66%
#### West	          0.50%	               25.10%

### Based on this values, I would suggest that there is NO difference in spending habits within the different regions of the U.S. 

## Question 4:
### The Instacart CFO isn’t interested in customers who don’t generate much revenue for the app. Create an exclusion flag for low-activity customers (customers with less than 5 orders) and exclude them from the data. Make sure you export this sample.

In [17]:
# Create exclusion flag to identify low-activity-customers
df_instacart.loc[df_instacart['max_order'] < 5, 'exclusion_flag'] = 'low_activity_customer'
df_instacart.loc[df_instacart['max_order'] >= 5, 'exclusion_flag'] = 'high_activity_customer'

In [18]:
# Counts on each regions:
df_instacart['exclusion_flag'].value_counts(dropna = False)

high_activity_customer    31929979
low_activity_customer       711289
Name: exclusion_flag, dtype: int64

In [19]:
# Check counts
31929979+711289

32641268

In [20]:
# Create the 2 lists according to customer-activity 
df_low_activity_customers = df_instacart[df_instacart['exclusion_flag'] == 'low_activity_customer']
df_high_activity_customers = df_instacart[df_instacart['exclusion_flag'] == 'high_activity_customer']

In [21]:
# Check maximum value of orders of 'low_activity_customers'. Must be below 5.
df_low_activity_customers['max_order'].value_counts()

4    711289
Name: max_order, dtype: int64

In [22]:
# Check maximum value of orders of 'high_activity_customers'. Must be greater than or equal to 5.
df_high_activity_customers['max_order'].value_counts().sort_index()

5       773857
6       810010
7       825922
8       816445
9       822821
        ...   
96       59965
97       40523
98       45019
99       44651
100    1173687
Name: max_order, Length: 96, dtype: int64

In [23]:
# Export both samples
df_low_activity_customers.to_pickle(os.path.join(path,'02 Data','Prepared Data','low_activity_customers.pkl'))
df_high_activity_customers.to_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_all.pkl'))

In [24]:
# Import data without the low-activity_customers as 
df_final = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_all.pkl'))

## Question 5: 
### The marketing and business strategy units at Instacart want to create more-relevant marketing strategies for different products and are, thus, curious about customer profiling in their database. Create a profiling variable based on age, income, certain goods in the “department_id” column, and number of dependents. You might also use the “orders_day_of_the_week” and “order_hour_of_day” columns if you can think of a way they would impact customer profiles. (Hint: As an example, try thinking of what characteristics would lead you to the profile “Single adult” or “Young parent.”)

In [25]:
# Let's take a look at the columns and info we have
df_final.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_label', 'Busy_days_by_orders',
       'Busiest_period_of_day', 'max_order', 'loyalty_flag', 'avg_price',
       'spending_flag', 'frequency_of_orders', 'order_frequency_flag',
       'gender', 'state', 'age', 'date_joined', 'num_dependents', 'fam_status',
       'income', 'region', 'exclusion_flag'],
      dtype='object')

In [26]:
# Descriptive Stats on numerical data
df_final.describe()

order_number  orders_day_of_week    order_time  days_since_prior_order  \
count  3.192998e+07        3.192998e+07  3.192998e+07            3.007955e+07   
mean   1.747409e+01        2.740343e+00  1.341951e+01            1.100350e+01   
std    1.757683e+01        2.089913e+00  4.247642e+00            8.711371e+00   
min    1.000000e+00        0.000000e+00  0.000000e+00            0.000000e+00   
25%    5.000000e+00        1.000000e+00  1.000000e+01            5.000000e+00   
50%    1.100000e+01        3.000000e+00  1.300000e+01            8.000000e+00   
75%    2.400000e+01        5.000000e+00  1.600000e+01            1.500000e+01   
max    1.000000e+02        6.000000e+00  2.300000e+01            3.000000e+01   

         product_id  add_to_cart_order     reordered      aisle_id  \
count  3.174776e+07       3.174776e+07  3.174776e+07  3.171813e+07   
mean   2.557630e+04       8.356248e+00  5.980359e-01  7.119179e+01   
std    1.409520e+04       7.130957e+00  4.902948e-01  3.821591e+01   
min    1.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00   
25%    1.353500e+04       3.000000e+00  0.000000e+00  3.100000e+01   
50%    2.525600e+04       6.000000e+00  1.000000e+00  8.300000e+01   
75%    3.793500e+04       1.100000e+01  1.000000e+00  1.070000e+02   
max    4.968800e+04       1.450000e+02  1.000000e+00  1.340000e+02   

       department_id        prices     max_order     avg_price  \
count   3.171813e+07  3.171813e+07  3.192998e+07  3.192998e+07   
mean    9.920806e+00  1.196038e+01  3.460900e+01  1.196174e+01   
std     6.281934e+00  4.939463e+02  2.502192e+01  8.204553e+01   
min     1.000000e+00  1.000000e+00  5.000000e+00  1.000000e+00   
25%     4.000000e+00  4.200000e+00  1.500000e+01  7.389583e+00   
50%     9.000000e+00  7.400000e+00  2.800000e+01  7.825000e+00   
75%     1.600000e+01  1.130000e+01  4.900000e+01  8.252408e+00   
max     2.100000e+01  9.999900e+04  1.000000e+02  1.500694e+04   

       frequency_of_orders           age  num_dependents        income  
count         3.192998e+07  3.192998e+07    3.192998e+07  3.192998e+07  
mean          1.022096e+01  4.946709e+01    1.501830e+00  9.953343e+04  
std           6.931400e+00  1.848690e+01    1.118833e+00  4.308692e+04  
min           0.000000e+00  1.800000e+01    0.000000e+00  2.590300e+04  
25%           6.000000e+00  3.300000e+01    1.000000e+00  6.712400e+04  
50%           8.000000e+00  4.900000e+01    2.000000e+00  9.667400e+04  
75%           1.300000e+01  6.500000e+01    3.000000e+00  1.279750e+05  
max           3.000000e+01  8.100000e+01    3.000000e+00  5.939010e+05

In [27]:
# Age: Create age groups. Customers vary between 18 to 81 years of age
df_final.loc[(df_final['age'] <= 24), 'age_group'] = '18 - 24'
df_final.loc[(df_final['age'] >= 25) & (df_final['age'] < 35), 'age_group'] = '25 - 34'
df_final.loc[(df_final['age'] >= 35) & (df_final['age'] < 45), 'age_group'] = '35 - 44'
df_final.loc[(df_final['age'] >= 45) & (df_final['age'] < 55), 'age_group'] = '45 - 54'
df_final.loc[(df_final['age'] >= 55) & (df_final['age'] < 65), 'age_group'] = '55 - 64'
df_final.loc[(df_final['age'] >= 65) & (df_final['age'] < 75), 'age_group'] = '65 - 74'
df_final.loc[(df_final['age'] >= 75) & (df_final['age'] < 85), 'age_group'] = '75 - 84'
df_final.loc[(df_final['age'] >= 85), 'age_group'] = '85+'

In [28]:
# Check counts
df_final['age_group'].value_counts().sort_index()

18 - 24    3522568
25 - 34    4977804
35 - 44    4987934
45 - 54    5037753
55 - 64    4953072
65 - 74    4930601
75 - 84    3520247
Name: age_group, dtype: int64

In [29]:
# Income: Create groups of income by $50,000. Values vary between $25,903 - $593,901
df_final.loc[(df_final['income'] < 50000), 'income_group'] = '< 50,000'
df_final.loc[(df_final['income'] >= 50000) & (df_final['income'] < 100000), 'income_group'] = '50,000 - 99,999'
df_final.loc[(df_final['income'] >= 100000) & (df_final['income'] < 150000), 'income_group'] = '100,000 - 149,999'
df_final.loc[(df_final['income'] >= 150000) & (df_final['income'] < 200000), 'income_group'] = '150,000 - 199,999'
df_final.loc[(df_final['income'] >= 200000) & (df_final['income'] < 250000), 'income_group'] = '200,000 - 249,999'
df_final.loc[(df_final['income'] >= 250000) & (df_final['income'] < 300000), 'income_group'] = '250,000 - 299,999'
df_final.loc[(df_final['income'] >= 300000) & (df_final['income'] < 350000), 'income_group'] = '300,000 - 349,999'
df_final.loc[(df_final['income'] >= 350000) & (df_final['income'] < 400000), 'income_group'] = '350,000 - 399,999'
df_final.loc[(df_final['income'] >= 400000) & (df_final['income'] < 450000), 'income_group'] = '400,000 - 449,999'
df_final.loc[(df_final['income'] >= 450000) & (df_final['income'] < 500000), 'income_group'] = '450,000 - 499,999'
df_final.loc[(df_final['income'] >= 500000) & (df_final['income'] < 550000), 'income_group'] = '500,000 - 549,999'
df_final.loc[(df_final['income'] >= 550000), 'income_group'] = '550,000+'

In [30]:
# Check counts
df_final['income_group'].value_counts().sort_index()

100,000 - 149,999    10618781
150,000 - 199,999     3743090
200,000 - 249,999       75596
250,000 - 299,999       80817
300,000 - 349,999       27339
350,000 - 399,999       26508
400,000 - 449,999       19280
450,000 - 499,999        9033
50,000 - 99,999      13799255
500,000 - 549,999        9477
550,000+                 9730
< 50,000              3511073
Name: income_group, dtype: int64

In [31]:
# Dependents: Parents or Not Parents according to the number of dependents. Values vary between 0 and 3.
df_final.loc[(df_final['num_dependents'] == 0), 'parental_status'] = 'Not Parent'
df_final.loc[(df_final['num_dependents'] >= 1), 'parental_status'] = 'Parent'

In [32]:
# Check the family_status categories
df_final['parental_status'].value_counts()

Parent        23950452
Not Parent     7979527
Name: parental_status, dtype: int64

In [33]:
# Order_time: Create groups of 4 hours in lenght to identify customer behavior
df_final.loc[(df_final['order_time'] >= 1) & (df_final['order_time'] <= 4), 'order_time_group'] = '1 - 4hrs (Late Night)'
df_final.loc[(df_final['order_time'] >= 5) & (df_final['order_time'] <= 8), 'order_time_group'] = '4 - 8hrs (Early Morning)'
df_final.loc[(df_final['order_time'] >= 9) & (df_final['order_time'] <= 12), 'order_time_group'] = '9- 12hrs (Late Morning)'
df_final.loc[(df_final['order_time'] >= 13) & (df_final['order_time'] <= 16), 'order_time_group'] = '13 - 16hrs (Afternoon)'
df_final.loc[(df_final['order_time'] >= 17) & (df_final['order_time'] <= 20), 'order_time_group'] = '17 - 20hrs (Evening)'
df_final.loc[(df_final['order_time'] >= 21) & (df_final['order_time'] <= 23) | (df_final['order_time'] == 0), 'order_time_group'] = '21 - 0hrs (Night)'

In [34]:
df_final['order_time_group'].value_counts()

9- 12hrs (Late Morning)     10422295
13 - 16hrs (Afternoon)      10383513
17 - 20hrs (Evening)         5858898
4 - 8hrs (Early Morning)     2959129
21 - 0hrs (Night)            2021264
1 - 4hrs (Late Night)         284880
Name: order_time_group, dtype: int64

In [35]:
# Day of the week: Create another column with the name of the day of the week
df_final.loc[df_final['orders_day_of_week'] == 0, 'weekday'] = 'Saturday'
df_final.loc[df_final['orders_day_of_week'] == 1, 'weekday'] = 'Sunday'
df_final.loc[df_final['orders_day_of_week'] == 2, 'weekday'] = 'Monday'
df_final.loc[df_final['orders_day_of_week'] == 3, 'weekday'] = 'Tuesday'
df_final.loc[df_final['orders_day_of_week'] == 4, 'weekday'] = 'Wednesday'
df_final.loc[df_final['orders_day_of_week'] == 5, 'weekday'] = 'Thursday'
df_final.loc[df_final['orders_day_of_week'] == 6, 'weekday'] = 'Friday'

In [36]:
# Before creating the profiles, I want to make sure tha t the department name appears. From a previuos script (4.10) I gor the description of each deparment. 
df_final.loc[df_final['department_id'] == 1, 'department_name'] = 'Frozen'
df_final.loc[df_final['department_id'] == 2, 'department_name'] = 'Other'
df_final.loc[df_final['department_id'] == 3, 'department_name'] = 'Bakery'
df_final.loc[df_final['department_id'] == 4, 'department_name'] = 'Produce'
df_final.loc[df_final['department_id'] == 5, 'department_name'] = 'Alcohol'
df_final.loc[df_final['department_id'] == 6, 'department_name'] = 'International'
df_final.loc[df_final['department_id'] == 7, 'department_name'] = 'Beverages'
df_final.loc[df_final['department_id'] == 8, 'department_name'] = 'Pets'
df_final.loc[df_final['department_id'] == 9, 'department_name'] = 'Dry goods pasta'
df_final.loc[df_final['department_id'] == 10, 'department_name'] = 'Bulk'
df_final.loc[df_final['department_id'] == 11, 'department_name'] = 'Personal care'
df_final.loc[df_final['department_id'] == 12, 'department_name'] = 'Meat seafood'
df_final.loc[df_final['department_id'] == 13, 'department_name'] = 'Pantry'
df_final.loc[df_final['department_id'] == 14, 'department_name'] = 'Breakfast'
df_final.loc[df_final['department_id'] == 15, 'department_name'] = 'Canned goods'
df_final.loc[df_final['department_id'] == 16, 'department_name'] = 'Dairy Eggs'
df_final.loc[df_final['department_id'] == 17, 'department_name'] = 'Household'
df_final.loc[df_final['department_id'] == 18, 'department_name'] = 'Babies'
df_final.loc[df_final['department_id'] == 19, 'department_name'] = 'Snacks'
df_final.loc[df_final['department_id'] == 20, 'department_name'] = 'Deli'
df_final.loc[df_final['department_id'] == 21, 'department_name'] = 'Missing'

In [37]:
# Want to get the counts for each department
df_final['department_name'].value_counts()

Produce            9288860
Dairy Eggs         5292745
Snacks             2829767
Beverages          2633114
Frozen             2180787
Pantry             1831123
Bakery             1147983
Canned goods       1041171
Deli               1028707
Dry goods pasta     845368
Household           720087
Meat seafood        692678
Breakfast           687659
Personal care       436395
Babies              417536
International       262920
Alcohol             149309
Pets                 95470
Missing              67034
Other                35403
Bulk                 34011
Name: department_name, dtype: int64

In [38]:
# Before continuing, I will change the type of the new variables to save some memory 
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31929979 entries, 0 to 32641267
Data columns (total 37 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                category
 1   user_id                 category
 2   order_number            int16   
 3   orders_day_of_week      int8    
 4   order_time              int8    
 5   days_since_prior_order  float64 
 6   product_id              float64 
 7   add_to_cart_order       float64 
 8   reordered               float64 
 9   product_name            category
 10  aisle_id                float64 
 11  department_id           float64 
 12  prices                  float64 
 13  price_label             category
 14  Busy_days_by_orders     category
 15  Busiest_period_of_day   category
 16  max_order               int16   
 17  loyalty_flag            category
 18  avg_price               float64 
 19  spending_flag           category
 20  frequency_of_orders     float64 
 21  order_

In [39]:
# I will change all the objects into category
df_final = df_final.astype({'region' : 'category',
                            'exclusion_flag' : 'category',
                            'age_group' : 'category',
                            'income_group' : 'category',
                            'parental_status' : 'category',
                            'order_time_group' : 'category',
                            'weekday' : 'category',
                            'department_name' : 'category'})

In [40]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31929979 entries, 0 to 32641267
Data columns (total 37 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                category
 1   user_id                 category
 2   order_number            int16   
 3   orders_day_of_week      int8    
 4   order_time              int8    
 5   days_since_prior_order  float64 
 6   product_id              float64 
 7   add_to_cart_order       float64 
 8   reordered               float64 
 9   product_name            category
 10  aisle_id                float64 
 11  department_id           float64 
 12  prices                  float64 
 13  price_label             category
 14  Busy_days_by_orders     category
 15  Busiest_period_of_day   category
 16  max_order               int16   
 17  loyalty_flag            category
 18  avg_price               float64 
 19  spending_flag           category
 20  frequency_of_orders     float64 
 21  order_

In [41]:
# Now get the mode of each age group, to use in the profiles
df = (df_final.groupby('age_group', as_index = False)
        .agg({'age':'mean', 'department_id': lambda x: x.value_counts().index[0]}))
print (df)

  age_group        age  department_id
0   18 - 24  21.009055            4.0
1   25 - 34  29.517397            4.0
2   35 - 44  39.510974            4.0
3   45 - 54  49.496779            4.0
4   55 - 64  59.490012            4.0
5   65 - 74  69.535526            4.0
6   75 - 84  78.007186            4.0


In [42]:
# Count number of male/female
df_final['gender'].value_counts()

Male      16077602
Female    15852377
Name: gender, dtype: int64

In [43]:
# Check that the 5 new columns are part of the dataframe (age_group, income_group, parental_status, order_time_group, department_name, and weekday)
pd.options.display.max_columns = None
df_final.head(1)

order_id user_id  order_number  orders_day_of_week  order_time  \
0  2539329       1             1                   2           8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda      77.0            7.0     9.0  Mid-range product   

  Busy_days_by_orders Busiest_period_of_day  max_order      loyalty_flag  \
0      Regularly days        Average orders         11  Regular customer   

   avg_price spending_flag  frequency_of_orders order_frequency_flag  gender  \
0   6.367797   Low spender                 20.0     Regular customer  Female   

     state  age date_joined  num_dependents fam_status  income region  \
0  Alabama   31   2/17/2019               3    married   40423  South   

           exclusion_flag age_group income_group parental_status  \
0  high_activity_customer   25 - 34     < 50,000          Parent   

           order_time_group weekday department_name  
0  4 - 8hrs (Early Morning)  Monday       Beverages

In [44]:
df_final.shape

(31929979, 37)

## Now, I will create some profiles according to parental status, age, income, goods in certain departments.

In [45]:
# Low income Parent: Parent, earns less than 50,000dlls, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli
df_final.loc[(df_final['parental_status'] == 'Parent') & (df_final['income'] < 50000) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'Low income parent'
# Young Parent: Parent, >50,000dlls, age 25 - 44, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli
df_final.loc[(df_final['parental_status'] == 'Parent') & (df_final['income'] >= 50000) & (df_final['income'] < 300000)& (df_final['age'] < 45) &  (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'Young parent'
# Adult Parent: Parent, >50,000, age 45 - 64, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli 
df_final.loc[(df_final['parental_status'] == 'Parent') & (df_final['income'] >= 50000) & (df_final['income'] < 300000)& (df_final['age'] >= 45) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'Adult parent'
# Older Parent: Parent, >50,000dlls, age 65+, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli
df_final.loc[(df_final['parental_status'] == 'Parent') & (df_final['income'] >= 50000) & (df_final['income'] < 300000) & (df_final['age'] >= 65) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'Older parent'
# High income Parent: Parent, earns more than 3000,000dlls, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli
df_final.loc[(df_final['parental_status'] == 'Parent') & (df_final['income'] >= 300000) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'High income parent'

# Low income single: Not-Parent, earns less than 50,000dlls, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli
df_final.loc[(df_final['parental_status'] == 'Not Parent') & (df_final['income'] < 50000) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'Low income single'
# Young single: Not-Parent, age 18 - 34, >50,000, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli
df_final.loc[(df_final['parental_status'] == 'Not Parent') &  (df_final['income'] >= 50000) & (df_final['income'] < 300000) & (df_final['age'] < 45) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'Young single'
# Adult single: Not-Parent, age 35 - 64yrs, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli
df_final.loc[(df_final['parental_status'] == 'Not Parent') & (df_final['income'] >= 50000) & (df_final['income'] < 300000) & (df_final['age'] >= 45) & (df_final['age'] < 65) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'Adult single'
# Older single: Not-parent, 65+ yrs 50,000 <= income > 300,000, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli 
df_final.loc[(df_final['parental_status'] == 'Not Parent') & (df_final['income'] >= 50000) & (df_final['income'] < 300000) & (df_final['age'] >= 65) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'Older single'
# High income single: Not-Parent, earns more than 3000,000dlls, buys produce or dairy/eggs or snacks or beverages or frozen or pantry or bakery or canned or deli
df_final.loc[(df_final['parental_status'] == 'Not Parent') & (df_final['income'] >= 300000) & (df_final['department_id'].isin([4,16,19,7,1,13,3,15,20])), 'profiling_variable'] = 'High income single'

In [46]:
# Check the counts of each profile
df_final['profiling_variable'].value_counts()

Young parent          7369160
Adult parent          5796516
Older parent          4947848
Young single          2435007
Low income parent     2277998
Adult single          1960739
Older single          1639805
Low income single      765752
High income parent      59471
High income single      21961
Name: profiling_variable, dtype: int64

In [47]:
# Quick check on ne columns
df_final.tail(10)

order_id user_id  order_number  orders_day_of_week  order_time  \
32641258  2977660  206209            13                   1          12   
32641259  2977660  206209            13                   1          12   
32641260  2977660  206209            13                   1          12   
32641261  2977660  206209            13                   1          12   
32641262  2977660  206209            13                   1          12   
32641263  2977660  206209            13                   1          12   
32641264  2977660  206209            13                   1          12   
32641265  2977660  206209            13                   1          12   
32641266  2977660  206209            13                   1          12   
32641267   272231  206209            14                   6          14   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
32641258                     7.0     24852.0                1.0        1.0   
32641259                     7.0      9405.0                2.0        1.0   
32641260                     7.0     16168.0                3.0        1.0   
32641261                     7.0     39216.0                4.0        0.0   
32641262                     7.0     14197.0                5.0        1.0   
32641263                     7.0     38730.0                6.0        0.0   
32641264                     7.0     31477.0                7.0        0.0   
32641265                     7.0      6567.0                8.0        0.0   
32641266                     7.0     22920.0                9.0        0.0   
32641267                    30.0         NaN                NaN        NaN   

                                              product_name  aisle_id  \
32641258                                            Banana      24.0   
32641259  Calcium Enriched 100% Lactose Free Fat Free Milk      91.0   
32641260                   Large Organic Omega3 Brown Eggs      86.0   
32641261              Original Multigrain Spoonfuls Cereal     121.0   
32641262                                      Tomato Paste       9.0   
32641263                   Brownie Crunch High Protein Bar       3.0   
32641264             High Protein Bar Chunky Peanut Butter       3.0   
32641265               Chocolate Peanut Butter Protein Bar       3.0   
32641266               Roasted & Salted Shelled Pistachios     117.0   
32641267                                               NaN       NaN   

          department_id  prices        price_label Busy_days_by_orders  \
32641258            4.0    12.3                NaN        Busiest days   
32641259           16.0    11.5                NaN        Busiest days   
32641260           16.0    11.8                NaN        Busiest days   
32641261           14.0     3.3  Low-range product        Busiest days   
32641262            9.0     5.6                NaN        Busiest days   
32641263           19.0     5.9                NaN        Busiest days   
32641264           19.0     4.2  Low-range product        Busiest days   
32641265           19.0     4.9  Low-range product        Busiest days   
32641266           19.0     1.7  Low-range product        Busiest days   
32641267            NaN     NaN                NaN      Regularly days   

         Busiest_period_of_day  max_order      loyalty_flag  avg_price  \
32641258           Most orders         14  Regular customer   7.058915   
32641259           Most orders         14  Regular customer   7.058915   
32641260           Most orders         14  Regular customer   7.058915   
32641261           Most orders         14  Regular customer   7.058915   
32641262           Most orders         14  Regular customer   7.058915   
32641263           Most orders         14  Regular customer   7.058915   
32641264           Most orders         14  Regular customer   7.058915   
32641265           Most orders         14  Regular customer   7.058915   
32641266           Most orders         14  Regular

In [48]:
# Impute 'Other' in nans
df_final['profiling_variable'].fillna('Other', inplace=True)

In [49]:
df_final['profiling_variable'].value_counts()

Young parent          7369160
Adult parent          5796516
Older parent          4947848
Other                 4655722
Young single          2435007
Low income parent     2277998
Adult single          1960739
Older single          1639805
Low income single      765752
High income parent      59471
High income single      21961
Name: profiling_variable, dtype: int64

In [50]:
# check
df_final.tail(10)

order_id user_id  order_number  orders_day_of_week  order_time  \
32641258  2977660  206209            13                   1          12   
32641259  2977660  206209            13                   1          12   
32641260  2977660  206209            13                   1          12   
32641261  2977660  206209            13                   1          12   
32641262  2977660  206209            13                   1          12   
32641263  2977660  206209            13                   1          12   
32641264  2977660  206209            13                   1          12   
32641265  2977660  206209            13                   1          12   
32641266  2977660  206209            13                   1          12   
32641267   272231  206209            14                   6          14   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
32641258                     7.0     24852.0                1.0        1.0   
32641259                     7.0      9405.0                2.0        1.0   
32641260                     7.0     16168.0                3.0        1.0   
32641261                     7.0     39216.0                4.0        0.0   
32641262                     7.0     14197.0                5.0        1.0   
32641263                     7.0     38730.0                6.0        0.0   
32641264                     7.0     31477.0                7.0        0.0   
32641265                     7.0      6567.0                8.0        0.0   
32641266                     7.0     22920.0                9.0        0.0   
32641267                    30.0         NaN                NaN        NaN   

                                              product_name  aisle_id  \
32641258                                            Banana      24.0   
32641259  Calcium Enriched 100% Lactose Free Fat Free Milk      91.0   
32641260                   Large Organic Omega3 Brown Eggs      86.0   
32641261              Original Multigrain Spoonfuls Cereal     121.0   
32641262                                      Tomato Paste       9.0   
32641263                   Brownie Crunch High Protein Bar       3.0   
32641264             High Protein Bar Chunky Peanut Butter       3.0   
32641265               Chocolate Peanut Butter Protein Bar       3.0   
32641266               Roasted & Salted Shelled Pistachios     117.0   
32641267                                               NaN       NaN   

          department_id  prices        price_label Busy_days_by_orders  \
32641258            4.0    12.3                NaN        Busiest days   
32641259           16.0    11.5                NaN        Busiest days   
32641260           16.0    11.8                NaN        Busiest days   
32641261           14.0     3.3  Low-range product        Busiest days   
32641262            9.0     5.6                NaN        Busiest days   
32641263           19.0     5.9                NaN        Busiest days   
32641264           19.0     4.2  Low-range product        Busiest days   
32641265           19.0     4.9  Low-range product        Busiest days   
32641266           19.0     1.7  Low-range product        Busiest days   
32641267            NaN     NaN                NaN      Regularly days   

         Busiest_period_of_day  max_order      loyalty_flag  avg_price  \
32641258           Most orders         14  Regular customer   7.058915   
32641259           Most orders         14  Regular customer   7.058915   
32641260           Most orders         14  Regular customer   7.058915   
32641261           Most orders         14  Regular customer   7.058915   
32641262           Most orders         14  Regular customer   7.058915   
32641263           Most orders         14  Regular customer   7.058915   
32641264           Most orders         14  Regular customer   7.058915   
32641265           Most orders         14  Regular customer   7.058915   
32641266           Most orders         14  Regular

In [51]:
# Export the data 
df_final.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'final_data.pkl'))

## Questions 6-10 are on the notebook "4.10 Excel Reporting PART 1 - Questions 6 -10"